### PR TITLE
Azure Backup for SQL Server in Azure Virtual Machines

### PR DESCRIPTION
**Pull request by nilaydshah**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

* Azure Backup for SQL Server in Azure Virtual Machines

### CODE REVIEW

File: 101-recovery-services-vm-workload-backup/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+      "type": "bool",
+      "defaultValue": false,
+      "metadata": {
+        "description": "Conditional parameter for New or Existing Backup Policy"
+      }
+    },
+    "workloadType": {
+      "type": "string",
+      "defaultValue": "SQLDataBase",
+      "allowedValues": [
```
Comment (bmoore-msft) [inline 101-recovery-services-vm-workload-backup/azuredeploy.json]: if there is only one allowedValue, why is this a parameter?

File: 101-recovery-services-vm-workload-backup/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        "SQLDataBase"
+      ],
+      "metadata": {
+        "description": "Workload type which is installed in VM and Pre-Registration Steps are performed"
+      }
+    },
+    "protectedItemType": {
+      "type": "string",
+      "defaultValue": "AzureVmWorkloadSQLDatabaseProtectedItem",
+      "allowedValues": [
```
Comment (bmoore-msft) [inline 101-recovery-services-vm-workload-backup/azuredeploy.json]: if there is only one allowedValue, why is this a parameter?

File: 101-recovery-services-vm-workload-backup/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+      "name": "[concat(parameters('vaultName'), '/', variables('backupFabric'), '/', variables('containerType'), ';compute;', parameters('vmResourceGroup'),';',parameters('vmName'))]",
+      "dependsOn": [
+        "[resourceId('Microsoft.RecoveryServices/vaults/', parameters('vaultName'))]"
+      ],
+      "properties":{
+        "containerType": "[variables('containerType')]",
+        "backupManagementType": "[variables('backupManagementType')]",
+        "workloadType": "[parameters('workloadType')]",
+        "friendlyName": "[parameters('vmName')]",
+        "sourceResourceId": "[resourceId(subscription().subscriptionId, parameters('vmResourceGroup'),'Microsoft.Compute/virtualMachines', parameters('vmName'))]"
```
Comment (bmoore-msft) [inline 101-recovery-services-vm-workload-backup/azuredeploy.json]: remove subscription().subscriptionID

File: 101-recovery-services-vm-workload-backup/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+          "isCompression": false
+        },
+        "subProtectionPolicy": [
+          {
+            "policyType": "Full",
+            "schedulePolicy": {
+                "schedulePolicyType": "SimpleSchedulePolicy",
+                "scheduleRunFrequency": "Daily",
+                "scheduleRunTimes": [
+                    "2019-03-24T16:00:00Z"
```
Comment (bmoore-msft) [inline 101-recovery-services-vm-workload-backup/azuredeploy.json]: what is this used for?

File: 101-recovery-services-vm-workload-backup/azuredeploy.parameters.json [hunk 5]
```diff
@@ -0,0 +1,27 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "vaultName": {
+            "value": "SqlBackupDemoVault"
```
Comment (bmoore-msft) [inline 101-recovery-services-vm-workload-backup/azuredeploy.parameters.json]: use GEN-UNIQUE - or add this as a defaultValue to the template and just remove it

File: 101-recovery-services-vm-workload-backup/azuredeploy.parameters.json [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        },
+        "vmName": {
+            "value": "GET-PREREQ-virtualMachineName"
+        },
+        "policyName": {
```
Comment (bmoore-msft) [inline 101-recovery-services-vm-workload-backup/azuredeploy.parameters.json]: remove these non-GEN values and use defaultValues in the template

File: 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+        "sqlAuthenticationPassword": {
+            "type": "securestring"
+        },
+        "rServicesEnabled": {
+            "type": "string"
+        }
+    },
+    "variables": {
+        "vnetId": "[resourceId(resourceGroup().name,'Microsoft.Network/virtualNetworks', parameters('virtualNetworkName'))]",
+        "subnetRef": "[concat(variables('vnetId'), '/subnets/', parameters('subnetName'))]"
```
Comment (bmoore-msft) [inline 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json]: remove vnetId and use

resourceId('Microsoft.Network/virtualNetworks/subnets', parameters('virtualNetworkName'), parameters('subnetName'))

File: 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+                },
+                "networkProfile": {
+                    "networkInterfaces": [
+                        {
+                            "id": "[resourceId('Microsoft.Network/networkInterfaces', parameters('networkInterfaceName'))]"
+                        }
+                    ]
+                }
+            },
+            "dependsOn": [
```
Comment (bmoore-msft) [inline 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json]: move this above properties

File: 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+                        "DiskCount": "[parameters('sqlStorageDisksCount')]",
+                        "DiskConfigurationType": "[parameters('sqlStorageDisksConfigurationType')]",
+                        "StartingDeviceID": "[parameters('sqlStorageStartingDeviceId')]"
+                    },
+                    "AdditionalFeaturesServerConfigurations": {
+                        "IsRServicesEnabled": "[parameters('rServicesEnabled')]"
+                    }
+                }
+            },
+            "dependsOn": [
```
Comment (bmoore-msft) [inline 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json]: move this above properties

File: 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+                "ipConfigurations": [
+                    {
+                        "name": "ipconfig1",
+                        "properties": {
+                            "subnet": {
+                                "id": "[variables('subnetRef')]"
+                            },
+                            "privateIPAllocationMethod": "Dynamic",
+                            "publicIpAddress": {
+                                "id": "[resourceId(resourceGroup().name,'Microsoft.Network/publicIpAddresses', parameters('publicIpAddressName'))]"
```
Comment (bmoore-msft) [inline 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json]: remove resourcegroup().name

File: 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+                            }
+                        }
+                    }
+                ],
+                "networkSecurityGroup": {
+                    "id": "[resourceId(resourceGroup().name, 'Microsoft.Network/networkSecurityGroups', parameters('networkSecurityGroupName'))]"
```
Comment (bmoore-msft) [inline 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json]: remove resourcegroup().name

File: 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+                }
+            },
+            "dependsOn": [
```
Comment (bmoore-msft) [inline 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json]: move this above properties

File: 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+            "properties": {
+                "securityRules": [
+                    {
+                        "name": "RDP",
+                        "properties": {
+                            "priority": 300,
+                            "protocol": "TCP",
+                            "access": "Allow",
+                            "direction": "Inbound",
+                            "sourceApplicationSecurityGroups": [],
```
Comment (bmoore-msft) [inline 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json]: remove empty arrays

File: 101-recovery-services-vm-workload-backup/azuredeploy.json (hunk 1 above)
Comment (nilaydshah) [reply 101-recovery-services-vm-workload-backup/azuredeploy.json]: planning to add support for more workloads as allowed values, like SAP HANA and ASE

File: 101-recovery-services-vm-workload-backup/azuredeploy.json (hunk 2 above)
Comment (nilaydshah) [reply 101-recovery-services-vm-workload-backup/azuredeploy.json]: planning to add support for more workloads as allowed values, like SAP HANA and ASE

File: 101-recovery-services-vm-workload-backup/azuredeploy.json (hunk 4 above)
Comment (nilaydshah) [reply 101-recovery-services-vm-workload-backup/azuredeploy.json]: This represents the time when scheduled backups are triggered

File: 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json (hunk 8 above)
Comment (nilaydshah) [reply 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json]: Done.

File: 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json (hunk 10 above)
Comment (nilaydshah) [reply 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json]: Done.

File: 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json (hunk 12 above)
Comment (nilaydshah) [reply 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json]: Done.

File: 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json (hunk 13 above)
Comment (nilaydshah) [reply 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json]: Done.

File: 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json (hunk 11 above)
Comment (nilaydshah) [reply 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json]: Done.

File: 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json (hunk 9 above)
Comment (nilaydshah) [reply 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json]: Done.

File: 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json (hunk 7 above)
Comment (nilaydshah) [reply 101-recovery-services-vm-workload-backup/prereqs/prereq.azuredeploy.json]: Done.

File: 101-recovery-services-vm-workload-backup/azuredeploy.parameters.json (hunk 6 above)
Comment (nilaydshah) [reply 101-recovery-services-vm-workload-backup/azuredeploy.parameters.json]: Done.

File: 101-recovery-services-vm-workload-backup/azuredeploy.parameters.json (hunk 5 above)
Comment (nilaydshah) [reply 101-recovery-services-vm-workload-backup/azuredeploy.parameters.json]: Done.

File: 101-recovery-services-vm-workload-backup/azuredeploy.json (hunk 3 above)
Comment (nilaydshah) [reply 101-recovery-services-vm-workload-backup/azuredeploy.json]: Done.

**Comment by nilaydshah** [comment]
@bmoore-msft  I've addressed all review comments.

File: 101-recovery-services-vm-workload-backup/azuredeploy.json [hunk 14]
```diff
@@ -82,6 +76,7 @@
     }
   },
   "variables": {
+    "vaultName": "SqlBackupDemoVault",
```
Comment (bmoore-msft) [inline 101-recovery-services-vm-workload-backup/azuredeploy.json]: revert this change - and add a defaultValue

File: 101-recovery-services-vm-workload-backup/azuredeploy.json (hunk 4 above)
Comment (bmoore-msft) [reply 101-recovery-services-vm-workload-backup/azuredeploy.json]: Should it be hardcoded or start when the template is deployed?

File: 101-recovery-services-vm-workload-backup/azuredeploy.parameters.json (hunk 6 above)
Comment (bmoore-msft) [reply 101-recovery-services-vm-workload-backup/azuredeploy.parameters.json]: I still see all the literal values here

File: 101-recovery-services-vm-workload-backup/azuredeploy.json (hunk 4 above)
Comment (nilaydshah) [reply 101-recovery-services-vm-workload-backup/azuredeploy.json]: This should be hardcoded, not dependent on time of deployment.

File: 101-recovery-services-vm-workload-backup/azuredeploy.parameters.json (hunk 6 above)
Comment (nilaydshah) [reply 101-recovery-services-vm-workload-backup/azuredeploy.parameters.json]: Oh yes, missed this reference. Fixed it now.

File: 101-recovery-services-vm-workload-backup/azuredeploy.json (hunk 14 above)
Comment (nilaydshah) [reply 101-recovery-services-vm-workload-backup/azuredeploy.json]: Done.

**Comment by bmoore-msft** [comment]
@nilaydshah - we're going to have to put these on hold.  The resources created by these templates (recovery services vaults) cannot be deleted - so I don't want to merge until that gets resolved.

**Comment by bmoore-msft** [comment]
@nilaydshah - we're going to have to put these on hold.  The resources created by these templates (recovery services vaults) cannot be deleted - so I don't want to merge until that gets resolved.

**Comment by nilaydshah** [comment]
@bmoore-msft  Vaults are deletable until it has not protected any backup item. And unless you remove those protected item first, as security reasons, we don't allow deletion of vaults. For this template, It's protecting SQL server database, unless we remove the protection (which is DELETE API call), we cannot delete the vault. So from backup side, it's by design and not a bug. 
What do you suggest in order to get this template merged?

File: 101-recovery-services-vm-workload-backup/azuredeploy.parameters.json (hunk 6 above)
Comment (bmoore-msft) [reply 101-recovery-services-vm-workload-backup/azuredeploy.parameters.json]: Still have the non-GEN values in here that need to be remoed

Comment (nilaydshah) [reply 101-recovery-services-vm-workload-backup/azuredeploy.parameters.json]: This is from older commit. Latest commit has below value
 "vmName": {
            "value": "sqlvmbackupdemo"
}

Comment (bmoore-msft) [reply 101-recovery-services-vm-workload-backup/azuredeploy.parameters.json]: What I see in the branch for this PR are non-GEN values.  These should be defaultValues in the template and removed from the parameters file.

**Comment by nilaydshah** [comment]
Oh I got it. 
I moved vmName from parameters file to default value in template. Is it recommended to remove all the below values as default values? In that case parameters file will be empty. Let me know your feedback.
        "policyName": {
            "value": "DailyFullHourlyLogDemo"
        },
        "databaseInstanceName": {
            "value": "mssqlserver"
        },
        "databaseName": {
            "value": "msdb"
        },
        "isNewVault": {
            "value": true
        },
        "isNewPolicy": {
            "value": true
        }
